### ISSUE TITLE
Re-ordering children in Visual Editor

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#350]
In a recent discussion Justin proposed a design plan to re-order a Visual Editor project's children/components in a _tree_ like manner. User can able to drag-n-drop any component into any group or containers within the tree hierarchy. Based on that the physical editor component positioning will also change. 

Upon more discussion on this with @JoelProminic , we agreed that when the tree can be easily made based on regular codes/components, some abstract controls like Grid/PanelGrid/TableView etc. which has column/cell nature, might take more to visually present to a tree component.

We agreed the priorities can be ordered in following way:

1. Tree representation of components
2. Ordering the children of a Div
3. Ordering children of a cell
4. Changing nesting from the table

**Comment by rat-moonshine** [comment]
We agreed that such tree component requires enough space vertically. Thus, the choice is to have a toggle system at sidebar, between property editor and the hierarchy editor (_tree_). You can propose otherwise, though.

**Comment by piotrzarzycki21** [comment]
What component for toggling are you going to use ? Accordion ?

**Comment by rat-moonshine** [comment]
No, you didn't get me. I meant to use complete height in the left panel between property and hierarchy editors. This can be done by some toggle buttons instead like this we can found in Microsoft Explorer (Windows 10):

![image](https://user-images.githubusercontent.com/24370870/44081767-28337ffa-9fcd-11e8-89d4-fc58a9d87dd7.png)

In my opinion, accordion usage would be more heavy and it'll complicate the natural views.

But anyway, this issue placed to lower priority. I shall work with you to complete the PanelGrid layout requisites and form generation first.

**Comment by piotrzarzycki21** [comment]
Looking into your last commit it looks like whole Tree will be placed in Moonshine not inside Visual Editor itself - Am I right ?

**Comment by rat-moonshine** [comment]
The re-ordering of editing components I don't think need to part of Moonshine codes but MockupVisualEditor. Thus, I thought to place the component at [this](https://github.com/prominic/MockupVisualEditor/blob/master/src/view/VisualEditor.mxml#L22) place, surrounded in a Container/Group and toggle system that can toggle between exiting PropertyEditorContainer and the re-ordering component.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine It would be great if you could next time make branches with features under folder **features**.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine I just build Visual Editor and Organizer displays for me like that 
![organizer](https://user-images.githubusercontent.com/24554795/45544078-a01f5780-b816-11e8-88a6-edd47bf9c189.png)

Probably something is missing.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 yes, thank you for mentioning this. The component is still incomplete. I was working on some (unfinish) UI styling - for that you also see the texts are white; There are many other things in to-dos, so you might face some unexpected behaviour by the organizer at this moment. I shall keep this updated as continue.

**Comment by JoelProminic** [comment]
I did some testing with this.   The dragging is working well for the most part, and it is easy to position a component as a child or sibling of panel.

Some general problems I noticed:
- Moving a control does not flag the editor tab as updated (*)
- Moves done in the Organizer tree cannot be undone

I noticed some issues with unsaved components:
1. Open a blank editor and add a Panel and a Button without saving
2. Open the Organizer.  
3. Drag the button to be a child of the new Div.  BUG:  in some cases, I got an error here.  However, it did not reproduce consistently.   If this did work, drag the Button back to the main Div for the next tests
4. In the mockup area, drag the button in the new Div.  BUG:  in the tree, the Button is not removed from the original position
5. In the mockup area, drag the button back to the main Div.  BUG:  in the tree, the Button is not emoved from the Div, and now appears twice at the base level.

I also got an error on the console when nesting a component under itself:
1. Add a panel component to the mockup area
2. Save
3. Drag the panel to be a child of itself

    : Error #1023
    : Error: Error #1023
    : 	at mx.core::UIComponent/get moduleFactory()
    : 	at mx.styles::StyleProtoChain$/getStyleManager()
    : 	at mx.styles::StyleProtoChain$/matchesCSSType()
    : 	at mx.core::UIComponent/matchesCSSType()
    : 	at mx.styles::CSSSelector/matchesStyleClient()
    : 	at mx.styles::CSSStyleDeclaration/matchesStyleClient()
    : 	at mx.styles::StyleProtoChain$/matchStyleDeclarations()
    : 	at mx.styles::StyleProtoChain$/getClassStyleDeclarations()
    : 	at mx.core::UIComponent/getClassStyleDeclarations()
    : 	at mx.styles::StyleProtoChain$/getMatchingStyleDeclarations()
    : 	at mx.styles::StyleProtoChain$/initProtoChain()
    : 	at mx.core::UIComponent/http://www.adobe.com/2006/flex/mx/internal::initProtoChain()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at spark.components::Group/regenerateStyleCache()
    : 	at mx.core::UIComponent/regenerateStyleCache()


This does not break Moonshine or the editor, though.  This action should not be allowed, but maybe we could display a different icon when the current drag position is an invalid move.

**Comment by rat-moonshine** [comment]
There some rules have been added so far, including one @JoelProminic you've pointed. 

- Tab, Grid, PanelGrid column/cells can't move within by dragging in Organizer
- Above components' column/cells also not movable outside of the origin component as well
- Parent component can't be drag into its children (one that @JoelProminic pointed in his last test)

**Comment by rat-moonshine** [comment]
Previously, 

- Highlight color changes to Organizer and resizer for better prominence
- Deleting in editor now updates to Organizer
- Deleting in Organizer now deletes in editor
- Selecting item in editor now highlight item in Organizer and auto-scroll
- Selecting item in organizer now selects item in editor (no auto-scroll)
- Drag-n-drop rules updated
- Organizer now auto-expand parents to depth and select non-displaying item

Today,

- Add/remove history management added to Organizer

**Comment by rat-moonshine** [comment]
As reported by @JoelProminic ,

> I found this issue:
> 
> 1. Add a Div
> 2. Add an OutputLabel inside the Div
> 3. Switch to Organizer
> 4. Undo twice. The OutputLabel and Div are removed as expected
> 5. Redo twice. BUG: both components reappear, but the OutputLabel appears at the top of the tree, outside of the Div

Fixed.

**Comment by rat-moonshine** [comment]
Children drag/move now added to the history management (_Note_: This applicable only to in-Organizer move).

**Comment by rat-moonshine** [comment]
Organizer needs to update itself on following property editor changes:

- [x] TabView tab addition
- [x] TabView tab deletion
- [ ] TabView tab title change
- [ ] PanelGrid row addition
- [x] PanelGrid row deletion
- [ ] PanelGrid column addition
- [x] PanelGrid column deletion
- [ ] Grid property editor changes as of PanelGrid's
- [ ] Undo management against above all

**Comment by rat-moonshine** [comment]
@JoelProminic reported following error to Organizer last night:

**Actual:**
1. Open a grid with two rows and columns
2. Select the first row
3. Add two columns.
4. Check Organizer. BUG: the new cells display after R2:C1
![image](https://user-images.githubusercontent.com/24370870/46802774-fe8b1780-cd7b-11e8-9c5a-9e02a687458f.png)

**Expected:**
R1:C1
R1:C2
R1:C3
R1:C4
R2:C1
R2:C2

Following things have covered in today's release:

- Root div is now visible in Organizer
- Above ordering problem reported by Joel should be fixed now
- Grid row/column generation through steppers now hooked to Organizer update
- PanelGrid row addition now updates to Organizer

**Comment by piotrzarzycki21** [comment]
I just make quick look into your changes:
1. Organizer displays name Stage - instead real name of component - I think it should be RootDiv or MainApplication
2. Organizer is displaying dash - is this have any purpose ? 
![organizer](https://user-images.githubusercontent.com/24554795/46817779-0056da00-cd80-11e8-815a-593c16d84a1e.png)

**Comment by rat-moonshine** [comment]
From a general user point of view, 'Stage' is much more simpler than technical terms like 'RootDiv' or 'MainApplication'. I'm keeping this 'Stage' terminology for now. Dash from the 'Organizer' label is removed. 

Following things also added with the current commit:

- Updating on PanelGrid column generation
- Adding a refresh button to manually refresh PanelGrid (in case some weird thing happen)
- Added disclaimer message (close-able) mentioning history manager is not available to the Organizer (for v1.16.0)
- Stopped in-editor component drag-move facility for PrimeFaces project (only) to reduce complexity (Organizer also do not listen to in-editor component moves)
- Organizer view will not avail for Visual Editor Flex project

**Comment by piotrzarzycki21** [comment]
@rat-moonshine Why you are not updating VisualEditor-0.2.0.swc along with commit to VisualEditor repository ?

**Comment by rat-moonshine** [comment]
I think I have done with the last piece of this implementation (unless some problem or unimplemented part again come up), and is ready for tests to go with upcoming release:

- Organizer now listen and updates upon Undo events to grid-cell/tabs/column changes through property editor and where it's applicable (i.e. Grid, TabView, PanelGrid etc.)
- Organizer node label now updates on _TabView_ _Tab_ label changes

Once we done with the tests, we can remove the current disclaimer message (to use Refresh button) from the interface. 

@Atinprominic, it'd be great if you have some time to test this as well.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine Are we ready to close that - Or is there any job to do ?
